### PR TITLE
Package builder fixes

### PR DESCRIPTION
**Pull request by aciidb0mb3r**

### CODE REVIEW

File: Tests/PackageLoadingTests/ConventionTests.swift [hunk 1]
```diff
@@ -463,16 +463,23 @@ class ConventionTests: XCTestCase {
         let fs = InMemoryFileSystem(emptyFiles:
             "/Sources/Foo/Foo.swift",
             "/Sources/Bar/Bar.swift",
-            "/Sources/Baz/Baz.swift")
+            "/Sources/Baz/Baz.swift",
+            "/Sources/Exec👽/main.swift"
+        )
```
Comment (ddunbar) [inline Tests/PackageLoadingTests/ConventionTests.swift]: Did you also make sure that the Build module won't do anything stupid with this (like try to link it in and fail)? We perhaps should have test coverage of that too.

Comment (aciidb0mb3r) [inline Tests/PackageLoadingTests/ConventionTests.swift]: Oh forgot about that, it does try to link. Will sep out this change from this PR.

**Comment by aciidb0mb3r** [comment]
@ddunbar updated

File: Sources/PackageLoading/PackageBuilder.swift [hunk 2]
```diff
@@ -433,6 +440,9 @@ public struct PackageBuilder {
                     if dependency.type != .library {
                         throw ModuleError.executableAsDependency(module: module.name, dependency: name)
                     }
+                    guard module.name != name else {
+                        throw ModuleError.cycleDetected("the target \(name) can't have a dependency on itself.")
+                    }
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageBuilder.swift]: This isn't actually checking for cycles, though?

My guess is this Package.swift blows up to:

targets: [Target(name: "A", dependencies: ["B"]), Target(name: "B", dependencies: "A"])

Comment (aciidb0mb3r) [inline Sources/PackageLoading/PackageBuilder.swift]: No this was just for checking self dependency right now, but I guess I can just do the cycles check instead.

Comment (ddunbar) [inline Sources/PackageLoading/PackageBuilder.swift]: We should, it's the more basic problem and will catch the same issues. We can add a new GraphAlgorithms method to find cycles in a graph (happens as a side effect of some of the existing ones, but we might want a standalone version if we want to detect it explicitly here).

**Comment by aciidb0mb3r** [comment]
Depends on #694

**Comment by aciidb0mb3r** [comment]
(This is updated and can be reviewed),

File: Sources/PackageLoading/PackageBuilder.swift [hunk 3]
```diff
@@ -47,6 +50,8 @@ extension ModuleError: FixableError {
             return "the target \(module) cannot have the executable \(dependency) as a dependency"
         case .invalidManifestConfig(let package, let message):
             return "invalid configuration in '\(package)': \(message)"
+        case .cycleDetected(let cycle):
+            return "found cyclic dependency declaration: " + cycle.map{$0.name}.joined(separator: " -> ")
```
Comment (ddunbar) [inline Sources/PackageLoading/PackageBuilder.swift]: I'd recommend reporting an extra `+ "-> \(cycle[0].name"` at the end to make it explicit where the cycle is.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Pull request merged**